File: tests/src/org.scala-lang/scala-library/2.11.8/src/test/scala/org_scala_lang/scala_library/EnumerationTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_scala_lang.scala_library

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import scala.Enumeration
import scala.collection.JavaConverters._

final class EnumerationTest {
  @Test
  def infersNamesForPublicValues(): Unit = {
    val names: Seq[String] = CardinalDirection.values.toSeq.map(_.toString)

    assertThat(names.asJava).containsExactly("North", "East", "South", "West")
    assertThat(CardinalDirection.withName("North")).isSameAs(CardinalDirection.North)
    assertThat(CardinalDirection.withName("West")).isSameAs(CardinalDirection.West)
  }

  @Test
  def deserializesValueToCanonicalEnumerationMember(): Unit = {
    val restored: CardinalDirection.Value = deserializeValue(serialize(CardinalDirection.East))

    assertThat(restored).isSameAs(CardinalDirection.East)
    assertThat(restored.toString).isEqualTo("East")
  }

  private def serialize(value: CardinalDirection.Value): Array[Byte] = {
    val bytes: ByteArrayOutputStream = new ByteArrayOutputStream()
    val out: ObjectOutputStream = new ObjectOutputStream(bytes)
    try {
      out.writeObject(value)
      out.flush()
      bytes.toByteArray
    } finally {
      out.close()
    }
  }

  private def deserializeValue(bytes: Array[Byte]): CardinalDirection.Value = {
    val in: ObjectInputStream = new ObjectInputStream(new ByteArrayInputStream(bytes))
    try {
      in.readObject().asInstanceOf[CardinalDirection.Value]
    } finally {
      in.close()
    }
  }
}

object CardinalDirection extends Enumeration {
  val North: Value = Value
  val East: Value = Value
  val South: Value = Value
  val West: Value = Value
}

File: tests/src/org.scala-lang/scala-library/2.11.8/src/test/java/org_scala_lang/scala_library/EnumerationTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_scala_lang.scala_library;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

import org.junit.jupiter.api.Test;

import scala.Enumeration;
import scala.math.BigDecimal.RoundingMode$;

import static org.assertj.core.api.Assertions.assertThat;

/** Exercises reflective value naming and singleton restoration. §FS-repository-functional-spec.5.2 */
public class EnumerationTest {

    @Test
    void resolvesValueNamesAndRestoresTheCanonicalEnumeration() throws Exception {
        RoundingMode$ roundingModes = RoundingMode$.MODULE$;

        Enumeration.Value halfEven = roundingModes.withName("HALF_EVEN");

        assertThat(halfEven).isSameAs(roundingModes.HALF_EVEN());
        assertThat(halfEven.toString()).isEqualTo("HALF_EVEN");
        assertThat(roundTrip(roundingModes)).isSameAs(roundingModes);
    }

    @SuppressWarnings("unchecked")
    private static <T> T roundTrip(T value) throws Exception {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        try (ObjectOutputStream output = new ObjectOutputStream(bytes)) {
            output.writeObject(value);
        }

        try (ObjectInputStream input = new ObjectInputStream(
                new ByteArrayInputStream(bytes.toByteArray()))) {
            return (T) input.readObject();
        }
    }
}
